stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/messages/adapter/viewholder/decorator/DeliveryStatusDecorator.kt

```diff
@@ -0,0 +1,92 @@
+import io.getstream.chat.android.ui.messages.adapter.viewholder.PlainTextWithFileAttachmentsViewHolder
+import io.getstream.chat.android.ui.messages.adapter.viewholder.PlainTextWithMediaAttachmentsViewHolder
+
+internal class DeliveryStatusDecorator : BaseDecorator() {
+
+    override fun decoratePlainTextMessage(viewHolder: MessagePlainTextViewHolder, data: MessageListItem.MessageItem) {
+        setupDeliveryStateIndicator(viewHolder.binding.deliveryStatusIcon, data)
+    }
+
+    override fun decoratePlainTextWithMediaAttachmentsMessage(
+        viewHolder: PlainTextWithMediaAttachmentsViewHolder,
+        data: MessageListItem.MessageItem
+    ) {
+        setupDeliveryStateIndicator(viewHolder.binding.deliveryStatusIcon, data)
+    }
+
+    override fun decorateOnlyMediaAttachmentsMessage(
+        viewHolder: OnlyMediaAttachmentsViewHolder,
+        data: MessageListItem.MessageItem
+    ) {
+        setupDeliveryStateIndicator(viewHolder.binding.deliveryStatusIcon, data)
+    }
+
+    override fun decorateOnlyFileAttachmentsMessage(
+        viewHolder: OnlyFileAttachmentsViewHolder,
+        data: MessageListItem.MessageItem
+    ) {
+        setupDeliveryStateIndicator(viewHolder.binding.deliveryStatusIcon, data)
+    }
+
+    override fun decoratePlainTextWithFileAttachmentsMessage(
+        viewHolder: PlainTextWithFileAttachmentsViewHolder,
+        data: MessageListItem.MessageItem
+    ) {
+        setupDeliveryStateIndicator(viewHolder.binding.deliveryStatusIcon, data)
+    }
+
+    override fun decorateDeletedMessage(viewHolder: MessageDeletedViewHolder, data: MessageListItem.MessageItem) = Unit
+
+    private fun setupDeliveryStateIndicator(imageView: ImageView, data: MessageListItem.MessageItem) {
+        fun hideIndicator() {
+            imageView.isVisible = false
+        }
+
+        fun showIndicator(@DrawableRes drawableRes: Int) {
+            imageView.isVisible = true
+            imageView.setImageResource(drawableRes)
+        }
+
+        // TODO review this logic for the BOTTOM check
+        if (BOTTOM !in data.positions) {
+            hideIndicator()
+            return
+        }
+
+        if (!data.isMine) {
+            hideIndicator()
+            return
+        }
+
+        when (data.message.syncStatus) {
+            SyncStatus.IN_PROGRESS, SyncStatus.SYNC_NEEDED -> {
+                showIndicator(R.drawable.stream_ui_ic_clock)
+            }
+            SyncStatus.COMPLETED -> {
+                if (data.messageReadBy.isNotEmpty()) {
+                    showIndicator(R.drawable.stream_ui_ic_check_all)
+                } else {
+                    showIndicator(R.drawable.stream_ui_ic_check_gray)
+                }
+            }
+            SyncStatus.FAILED_PERMANENTLY -> {
+                // This case is covered by a separate Decorator
+                hideIndicator()
+            }
+        }.exhaustive
```

Comment:
Why do we need `exhaustive` here?

Reply:
It just makes sure that all values of the enum are covered.